File: third/depend/MVS/Development/Samples/Java/Events_Interface/src/Events_Interface.java
/***************************************************************************************************
 * @file      Events_Interface.java
 * @breif     This demo show how to switch interface events
 * @author    zhanglei72
 * @date      2020/02/10
 *
 * @warning
 * @version   V1.0.0  2020/02/10 create this file.
 * @since     2020/02/10
 **************************************************************************************************/

import java.io.*;
import java.util.ArrayList;
import java.util.Scanner;

import MvCameraControlWrapper.*;
import static MvCameraControlWrapper.MvCameraControlDefines.*;

public class Events_Interface
{
	private static Handle _ifHandle = null;
    private static Handle _hCamera = null;
    private static volatile Boolean _grabThreadExit = false;
    private static Scanner _scanner = null;
	
	
    public static void main(String[] args)
    {
      
        int interfaceType = MV_GIGE_INTERFACE | MV_CAMERALINK_INTERFACE | MV_CXP_INTERFACE | MV_XOF_INTERFACE;
        _scanner = new Scanner(System.in);

		System.out.println("SDK Version " + MvCameraControl.MV_CC_GetSDKVersion());
            
		// Initialize SDK
		MvCameraControl.MV_CC_Initialize();
		
		ArrayList<MV_INTERFACE_INFO> stInterfaceList = null;
		int nEnumInterfaceType = -1;
        try
        {
				// Enumerate interfaces(frame grabbers)
				ArrayList<MV_INTERFACE_INFO> interfaceList = MvCameraControl.MV_CC_EnumInterfaces(interfaceType);
				if (interfaceList.size() == 0)
				{
					System.out.println("no interface found!");
					return;
				}

				printInterfaceInfo(interfaceList);

				int index = chooseInterface(interfaceList);
				if (index < 0 || index > interfaceList.size())
				{
					System.out.println("index is invalid!");
					return;
				}

				//Create interface handle
				_ifHandle = MvCameraControl.MV_CC_CreateInterface(interfaceList.get(index));

				//open interface
				int nRet = MvCameraControl.MV_CC_OpenInterface(_ifHandle, null);
				if (nRet != MV_OK)
				{
					System.out.printf("MV_CC_OpenInterface failed! nRet[%#x]\n", nRet);
					return;
				}
			
				// Turn on the specific event of interface
				nRet = MvCameraControl.MV_CC_EventNotificationOn(_ifHandle,"ReceiveImageFrameStart0");
				if (nRet != MV_OK)
				{
					System.err.printf("Event Notification On failed,nRet:[%#x]", nRet);
					return;
				}
				
				// Register image callback
            nRet = MvCameraControl.MV_CC_RegisterEventCallBack(_ifHandle, "ReceiveImageFrameStart0" ,new CameraEventCallBack() {
                @Override
                public int OnEventCallBack(MV_EVENT_OUT_INFO  info)
                {
                    printEventInfo(info);
                    return 0;
                }
            });
			
			
				//Enumerate device on the frame grabber
				ArrayList<MV_CC_DEVICE_INFO> deviceList = MvCameraControl.MV_CC_EnumDevicesByInterface(_ifHandle);
				if (deviceList.size() == 0)
				{
					System.out.println("no device found!");
					return;
				}

				printDeviceInfo(deviceList);

				// choose camera
				int camIndex = chooseCamera(deviceList);
				if (-1 == camIndex)
				{
					return;
				}

				// Create device handle
				try
				{
					_hCamera = MvCameraControl.MV_CC_CreateHandle(deviceList.get(camIndex));
				}
				catch (CameraControlException e)
				{
					System.err.println("Create handle failed!" + e.toString());
					e.printStackTrace();
					return;
				}

				// Open selected device
				nRet = MvCameraControl.MV_CC_OpenDevice(_hCamera);
				if (MV_OK != nRet)
				{
					System.err.printf("Connect to camera failed, errcode: [%#x]\n",nRet);
					return;
				}
				else
				{
					System.err.printf("Connect suc.\n");
				}


				// Turn on trigger mode
				nRet = MvCameraControl.MV_CC_SetEnumValueByString(_hCamera, "TriggerMode", "Off");
				if (MV_OK != nRet)
				{
					System.err.printf("SetTriggerMode failed, errcode: [%#x]\n", nRet);
					return;
				}

				// Start grabbing
				nRet = MvCameraControl.MV_CC_StartGrabbing(_hCamera);
				if (MV_OK != nRet)
				{
					System.err.printf("StartGrabbing failed, errcode: [%#x]\n", nRet);
					return;
				}
				else
				{
					System.err.printf("StartGrabbing  suc.\n");
				}


				_grabThreadExit = false;
				Thread thread = new Thread(new Runnable(){
					@Override
					public void run(){
                    // 在线程中执行函数
                    MV_FRAME_OUT stFrameOut = new MV_FRAME_OUT();
                    while(!_grabThreadExit)
                    {
                        int nRet = MvCameraControl.MV_CC_GetImageBuffer(_hCamera, stFrameOut, 1000);
                        // ch:获取一帧图像 | en:Get one image
                        if (MV_OK == nRet)
                        {
                          /*  System.out.println("Get Image Buffer Width:" + stFrameOut.mvFrameOutInfo.ExtendWidth
                                    + " Height: "   + stFrameOut.mvFrameOutInfo.ExtendHeight
                                    + " FrameNum: " + stFrameOut.mvFrameOutInfo.frameNum);

							*/
                            nRet = MvCameraControl.MV_CC_FreeImageBuffer(_hCamera, stFrameOut);
                            if (MV_OK != nRet)
                            {
                                System.err.printf("\n Free ImageBuffer failed, errcode: [%#x]\n", nRet);
                            }
                        }
                    }
                }
            });

            thread.start();

            _scanner.useDelimiter("");
            System.out.println("Press Enter to exit.");

            while(true)
            {
                String input = _scanner.nextLine();
                if(_scanner.hasNextLine())
                {
                    break;
                }
                else
                {
                    try {
                        Thread.sleep(1 * 10);
                    }
                    catch (InterruptedException e)
                    {
                        e.printStackTrace();
                        break;
                    }
                }
            }

            _scanner.close();

            _grabThreadExit = true;
            try {
                thread.join();
            } catch (InterruptedException e) {
                e.printStackTrace();
            }

            // Stop grabbing
            nRet = MvCameraControl.MV_CC_StopGrabbing(_hCamera);
            if (MV_OK != nRet)
            {
                System.err.printf("StopGrabbing failed, errcode: [%#x]\n", nRet);
                return;
            }

            // close device
            nRet = MvCameraControl.MV_CC_CloseDevice(_hCamera);
            if (MV_OK != nRet)
            {
                System.err.printf("CloseDevice failed, errcode: [%#x]\n", nRet);
                return;
            }
        }
		catch(CameraControlException e)
        {
            System.err.println("Enumrate interface failed!" + e.toString());
            e.printStackTrace();
        }
        finally
        {
            //关闭并销毁采集卡
            if (_ifHandle != null)
            {
                MvCameraControl.MV_CC_CloseInterface(_ifHandle);
                MvCameraControl.MV_CC_DestroyInterface(_ifHandle);
                _ifHandle = null;
            }
        }
       
		MvCameraControl.MV_CC_Finalize();
    }
	
	private static void printInterfaceInfo(ArrayList<MV_INTERFACE_INFO> interfaceList)
    {
        int i = 0;
        for (MV_INTERFACE_INFO stInterfaceInfo : interfaceList)
        {
            if (null == stInterfaceInfo)
            {
                continue;
            }
            System.out.println("[Interface " + (i++) + "]");

            System.out.println("\tTLayerType: " + stInterfaceInfo.TLayerType);
            System.out.println("\tPCIEInfo: " + stInterfaceInfo.PCIEInfo);
            System.out.println("\tSerial number: " + stInterfaceInfo.serialNumber);
            System.out.println("\tmodel name: " + stInterfaceInfo.modelName);
            System.out.println("\tInterfaceID: " + stInterfaceInfo.InterfaceID);
            System.out.println("\n");
        }
    }

	
    private  static void printDeviceInfo(ArrayList<MV_CC_DEVICE_INFO> deviceInfoList)
    {
        int i = 0;
        for (MV_CC_DEVICE_INFO deviceInfo : deviceInfoList)
        {
            if (null == deviceInfo)
            {
                continue;
            }
            System.out.println("[camera " + (i++) + "]");

            if (deviceInfo.transportLayerType == MV_GIGE_DEVICE || deviceInfo.transportLayerType == MV_GENTL_GIGE_DEVICE)
            {
                System.out.println("\tCurrentIp:       " + deviceInfo.gigEInfo.currentIp);
                System.out.println("\tModel:           " + deviceInfo.gigEInfo.modelName);
                System.out.println("\tUserDefinedName: " + deviceInfo.gigEInfo.userDefinedName);
            }
            else if (deviceInfo.transportLayerType == MV_USB_DEVICE)
            {
                System.out.println("\tUserDefinedName: " + deviceInfo.usb3VInfo.userDefinedName);
                System.out.println("\tSerial Number:   " + deviceInfo.usb3VInfo.serialNumber);
                System.out.println("\tDevice Number:   " + deviceInfo.usb3VInfo.deviceNumber);
            }
            else if (deviceInfo.transportLayerType == MV_GENTL_CAMERALINK_DEVICE)
            {
                System.out.println("\tUserDefinedName: " + deviceInfo.cmlInfo.userDefinedName);
                System.out.println("\tSerial Number:   " + deviceInfo.cmlInfo.serialNumber);
                System.out.println("\tDevice Number:   " + deviceInfo.cmlInfo.DeviceID);
            }
            else if (deviceInfo.transportLayerType == MV_GENTL_CXP_DEVICE)
            {
                System.out.println("\tUserDefinedName: " + deviceInfo.cxpInfo.userDefinedName);
                System.out.println("\tSerial Number:   " + deviceInfo.cxpInfo.serialNumber);
                System.out.println("\tDevice Number:   " + deviceInfo.cxpInfo.DeviceID);
            }
            else if (deviceInfo.transportLayerType == MV_GENTL_XOF_DEVICE)
            {
                System.out.println("\tUserDefinedName: " + deviceInfo.xofInfo.userDefinedName);
                System.out.println("\tSerial Number:   " + deviceInfo.xofInfo.serialNumber);
                System.out.println("\tDevice Number:   " + deviceInfo.xofInfo.DeviceID);
            }
            else
            {
                System.out.println("\tDevice is not supported! TransferType: " + deviceInfo.transportLayerType);
            }
        }
    }

   
   public static int chooseCamera(ArrayList<MV_CC_DEVICE_INFO> stDeviceList)
    {
        if (null == stDeviceList)
        {
            return -1;
        }
        
        // Choose a device to operate
        int camIndex = -1;

        while (true)
        {
			System.out.print("Please input camera index:");
			if (_scanner.hasNextInt()) 
			{
                try
                {
				   camIndex = _scanner.nextInt();
                   if ((camIndex >= 0 && camIndex < stDeviceList.size()) || -1 == camIndex)
                   {
                       break;
                   }
                   else
                  {
                      System.out.println("Input error: " + camIndex + " Over Range:( 0 - " + (stDeviceList.size()-1) + " )");
                   }
                }
                catch (Exception e)
                {
			       System.out.println("Input not number.");
                   camIndex = -1;
                   break;
                }
			}
			else
			{
				camIndex = -1;
                break;
			}
			
        }
       if (-1 == camIndex) {
            System.out.println("Input error.exit");
            return camIndex;
        }

        if (0 <= camIndex && stDeviceList.size() > camIndex)
        {
			//deviceInfo.transportLayerType == MV_GIGE_DEVICE || deviceInfo.transportLayerType == MV_GENTL_GIGE_DEVICE
            if ((MV_GIGE_DEVICE == stDeviceList.get(camIndex).transportLayerType)||(MV_GENTL_GIGE_DEVICE == stDeviceList.get(camIndex).transportLayerType))
            {
                System.out.println("Connect to camera[" + camIndex + "]: " + stDeviceList.get(camIndex).gigEInfo.userDefinedName);
            }
            else if (MV_USB_DEVICE == stDeviceList.get(camIndex).transportLayerType)
            {
                System.out.println("Connect to camera[" + camIndex + "]: " + stDeviceList.get(camIndex).usb3VInfo.userDefinedName);
            }
			 else if (stDeviceList.get(camIndex).transportLayerType == MV_GENTL_CAMERALINK_DEVICE)
            {
				System.out.println("Connect to camera[" + camIndex + "]: " + stDeviceList.get(camIndex).cmlInfo.userDefinedName);
            }
            else if (stDeviceList.get(camIndex).transportLayerType == MV_GENTL_CXP_DEVICE)
            {
				System.out.println("Connect to camera[" + camIndex + "]: " + stDeviceList.get(camIndex).cxpInfo.userDefinedName);
            }
            else if (stDeviceList.get(camIndex).transportLayerType == MV_GENTL_XOF_DEVICE)
            {
				System.out.println("Connect to camera[" + camIndex + "]: " + stDeviceList.get(camIndex).xofInfo.userDefinedName);
            }
            else
            {
                System.out.println("Device is not supported.TransferType: " + stDeviceList.get(camIndex).transportLayerType);
            }
        }
        else
        {
            System.out.println("Invalid index " + camIndex);
            camIndex = -1;
        }

        return camIndex;
    }
	
    public static int chooseInterface(ArrayList<MV_INTERFACE_INFO> stInterfaceList)
    {
        if (null == stInterfaceList)
        {
            return -1;
        }
        
        // Choose a Interface to operate
        int InterfaceIndex = -1;
       
        while (true)
        {
			System.out.print("Please input interface index (-1 to quit):");
			if (_scanner.hasNextInt()) 
			{
				try
                {
					InterfaceIndex = _scanner.nextInt();
					if ((InterfaceIndex >= 0 && InterfaceIndex < stInterfaceList.size()) || -1 == InterfaceIndex)
					{
						break;
					}
					else
					{
						System.out.println("Input error: " + InterfaceIndex);
					}
				}
				catch (Exception e)
				{
					e.printStackTrace();
					InterfaceIndex = -1;
					break;
				}
			}
			else{
				InterfaceIndex = -1;
				break;
			}
           
        }
      
        if (-1 == InterfaceIndex)
        {
            System.out.println("Input error.exit");
            return InterfaceIndex;
        }

        if (0 <= InterfaceIndex && stInterfaceList.size() > InterfaceIndex)
        {
            if (MV_GIGE_INTERFACE == stInterfaceList.get(InterfaceIndex).TLayerType)
            {
                System.out.println("Connect to Interface[" + InterfaceIndex + "]: " + stInterfaceList.get(InterfaceIndex).serialNumber);
            }
            else if (MV_CAMERALINK_INTERFACE == stInterfaceList.get(InterfaceIndex).TLayerType)
            {
                System.out.println("Connect to Interface[" + InterfaceIndex + "]: " + stInterfaceList.get(InterfaceIndex).serialNumber);
            }
			else if (MV_CXP_INTERFACE == stInterfaceList.get(InterfaceIndex).TLayerType)
            {
                System.out.println("Connect to Interface[" + InterfaceIndex + "]: " + stInterfaceList.get(InterfaceIndex).serialNumber);
            }
			else if (MV_XOF_INTERFACE == stInterfaceList.get(InterfaceIndex).TLayerType)
            {
                System.out.println("Connect to Interface[" + InterfaceIndex + "]: " + stInterfaceList.get(InterfaceIndex).serialNumber);
            }
            else
            {
                System.out.println("Interface is not supported.");
            }
        }
        else
        {
            System.out.println("Invalid index " + InterfaceIndex);
            InterfaceIndex = -1;
        }

        return InterfaceIndex;
    }
	
	public static void printEventInfo(MV_EVENT_OUT_INFO pEventInfo)
	{
		
		long nBlockId = pEventInfo.blockIdHigh;
		nBlockId = (nBlockId << 32) + pEventInfo.blockIdLow;

		long nTimestamp = pEventInfo.timestampHigh;
		nTimestamp = (nTimestamp << 32) + pEventInfo.timestampLow;

		System.err.printf("EventName[%s], EventID[%d], BlockId[%d], Timestamp[%d]\n", 
			pEventInfo.eventName, pEventInfo.eventID,nBlockId,nTimestamp);
		
	}


}

File: third/depend/MVS/Development/Samples/VC/VS/SimpleSamples/Events_Interface/Events_Interface.cpp
/*
 * 这个示例演示了打开并接收采集卡事件的功能
 * This sample shows how to turn on the specified event of frame grabber and receive the event.
 * 
 */
#include <stdio.h>
#include <Windows.h>
#include <process.h>
#include <conio.h>
#include "MvCameraControl.h"


//ch：打印设备信息 | en: Print the information of devices
bool PrintDeviceInfo(MV_CC_DEVICE_INFO_LIST& stDeviceList);
//ch：打印采集卡信息 | en: Print the information of frame grabbers
bool PrintInterfaceInfo(MV_INTERFACE_INFO_LIST& stInterfaceInfoList);
//ch: 事件回调函数 | en: Event callback function
void __stdcall EventCallBack(MV_EVENT_OUT_INFO * pEventInfo, void* pUser);
//ch: 取图回调函数 | en: Image callback function
void __stdcall ImageCallBackEx(unsigned char * pData, MV_FRAME_OUT_INFO_EX* pFrameInfo, void* pUser);

// ch:等待按键输入 | en:Wait for key press
void WaitForKeyPress(void)
{
	while(!_kbhit())
	{
		Sleep(10);
	}
	_getch();
}


int main()
{
	int nRet = MV_OK;

	//ch: 采集卡接口类型 | en: The interface type of frame grabber
	int nInterfaceType = MV_GIGE_INTERFACE | MV_CAMERALINK_INTERFACE | MV_CXP_INTERFACE | MV_XOF_INTERFACE;

	void* hInterface = NULL;  //ch: 采集卡句柄 | en: Handle of frame grabber
	void* hDevHandle = NULL;  //ch: 设备句柄 | en: Handle of device

	do 
	{
		// ch:初始化SDK | en:Initialize SDK
		nRet = MV_CC_Initialize();
		if (MV_OK != nRet)
		{
			printf("Initialize SDK fail! nRet [0x%x]\n", nRet);
			break;
		}

		//ch: 枚举采集卡 | en: Enumerate frame grabbers
		MV_INTERFACE_INFO_LIST stInterfaceInfoList={0};
		nRet = MV_CC_EnumInterfaces(nInterfaceType, &stInterfaceInfoList);
		if (MV_OK != nRet)
		{
			printf("Enum interfaces fail! nRet [0x%x]\n", nRet);
			break;
		}

		if (stInterfaceInfoList.nInterfaceNum > 0)
		{
			PrintInterfaceInfo(stInterfaceInfoList);
			printf("Enum interfaces success!\n\n");
		} 
		else
		{
			printf("Find no interface!\n");
			break;
		}

		printf("Please input interfaces index(0-%d):", stInterfaceInfoList.nInterfaceNum-1);
		unsigned int nIndex = 0;
		scanf_s("%d", &nIndex);

		if (nIndex >= stInterfaceInfoList.nInterfaceNum)
		{
			printf("Input error!\n");
			break;
		}

		//ch:创建采集卡句柄 | en: Create the handle of frame grabber
		nRet = MV_CC_CreateInterface(&hInterface, stInterfaceInfoList.pInterfaceInfos[nIndex]);
		if (MV_OK == nRet)
		{
			printf("Create interface success!\n");
		}
		else
		{
			printf("Create interface Handle fail! nRet [0x%x]\n", nRet);
			break;
		}

		//ch: 打开采集卡 | en: Open the frame grabber 
		nRet = MV_CC_OpenInterface(hInterface, NULL);
		if (MV_OK == nRet)
		{
			printf("Open interface success!\n");
		}
		else
		{
			printf("Open interface fail! nRet [0x%x]\n", nRet);
			break;
		}

        //ch：开启采集卡ReceiveImageFrameStart0事件 | en: Turn on the ReceiveImageFrameStart0 event of frame grabber
        nRet = MV_CC_EventNotificationOn(hInterface, "ReceiveImageFrameStart0");
		if (nRet != MV_OK)
		{
			printf("EventNotificationOn fail! nRet [0x%x]\n", nRet);
			break;
		}

		// ch:注册采集卡事件回调 | en:Register event callback
		nRet = MV_CC_RegisterEventCallBackEx(hInterface, "ReceiveImageFrameStart0", EventCallBack, NULL);
		if (MV_OK != nRet)
		{
			printf("Register Event CallBack fail! nRet [0x%x]\n", nRet);
			break; 
		}


		//ch: 枚举采集卡上的相机 | en: Enumerate the devices on the frame grabber
		MV_CC_DEVICE_INFO_LIST stDeviceList = {0};
		nRet = MV_CC_EnumDevicesByInterface(hInterface, &stDeviceList);
		if (nRet != MV_OK)
		{
			printf("MV_CC_EnumDevicesByInterface fail! nRet [0x%x]\n", nRet);
			break;
		}

		if (stDeviceList.nDeviceNum == 0)
		{
			printf("no device! fail!\n");
			break;
		}

		PrintDeviceInfo(stDeviceList);

		
		// ch:选择设备0并创建句柄 | en:Select device 0 and create handle
		nIndex = 0;
		nRet = MV_CC_CreateHandle(&hDevHandle, stDeviceList.pDeviceInfo[nIndex]);
		if (MV_OK != nRet)
		{
			printf("Create handle fail! nRet [0x%x]\n", nRet);
			break;
		}

		// ch:打开设备 | en:Open device
		nRet = MV_CC_OpenDevice(hDevHandle);
		if (MV_OK != nRet)
		{
			printf("Open device fail! nRet [0x%x]\n", nRet);
			break;
		}

		printf("Open device success!\n");

		// ch:注册取图回调 | en:Register image callback
		nRet = MV_CC_RegisterImageCallBackEx(hDevHandle, ImageCallBackEx, NULL);
		if (MV_OK != nRet)
		{
			printf("Register Image CallBack fail! nRet [0x%x]\n", nRet);
			break;
		}

		// ch:开始取流 | en:Start grab image
		nRet = MV_CC_StartGrabbing(hDevHandle);
		if (MV_OK != nRet)
		{
			printf("Start Grabbing fail! nRet [0x%x]\n", nRet);
			break;
		}

		printf("Press a key to stop grabbing.\n");
		WaitForKeyPress();

		// ch:停止取流 | en:Stop grab image
		nRet = MV_CC_StopGrabbing(hDevHandle);
		if (MV_OK != nRet)
		{
			printf("Stop Grabbing fail! nRet [0x%x]\n", nRet);
			break;
		}

		// ch:关闭设备 | Close device
		nRet = MV_CC_CloseDevice(hDevHandle);
		if (MV_OK != nRet)
		{
			printf("ClosDevice fail! nRet [0x%x]\n", nRet);
			break;
		}

		printf("Close device success!\n");

		// ch:销毁设备句柄 | Destroy handle of device
		nRet = MV_CC_DestroyHandle(hDevHandle);
		if (MV_OK != nRet)
		{
			printf("Destroy handle fail! nRet [0x%x]\n", nRet);
			break;
		}

		printf("Destroy device handle success!\n");

		hDevHandle = NULL;

		//ch: 关闭采集卡 | en: Close the frame grabber
		nRet = MV_CC_CloseInterface(hInterface);
		if (MV_OK == nRet)
		{
			printf("Close interface success!\n");
		}
		else
		{
			printf("Close interface Handle fail! nRet [0x%x]\n", nRet);
			break;
		}

		// ch:销毁采集卡句柄 | Destroy handle of frame grabber
		nRet = MV_CC_DestroyInterface(hInterface);
		if (MV_OK == nRet)
		{
			printf("Destroy interface success!\n");
		}
		else
		{
			printf("Destroy interface Handle fail! nRet [0x%x]\n", nRet);
			break;
		}
		hInterface = NULL;
	} while (0);

	if (hDevHandle != NULL)
	{
		MV_CC_CloseDevice(hDevHandle);
		MV_CC_DestroyHandle(hDevHandle);
		hDevHandle = NULL;
	}

	if (hInterface != NULL)
	{
		MV_CC_CloseInterface(hInterface);
		MV_CC_DestroyInterface(hInterface);
		hInterface = NULL;
	}

	// ch:反初始化SDK | en:Finalize SDK
	MV_CC_Finalize();

	printf("Press a key to exit.\n");
	WaitForKeyPress();
}

bool PrintDeviceInfo(MV_CC_DEVICE_INFO_LIST& stDeviceList)
{
	for (unsigned int i = 0; i < stDeviceList.nDeviceNum; i++)
	{
		printf("[device %d]:\n", i);
		MV_CC_DEVICE_INFO* pstMVDevInfo = stDeviceList.pDeviceInfo[i];

		if (pstMVDevInfo->nTLayerType == MV_GIGE_DEVICE)
		{
			int nIp1 = ((pstMVDevInfo->SpecialInfo.stGigEInfo.nCurrentIp & 0xff000000) >> 24);
			int nIp2 = ((pstMVDevInfo->SpecialInfo.stGigEInfo.nCurrentIp & 0x00ff0000) >> 16);
			int nIp3 = ((pstMVDevInfo->SpecialInfo.stGigEInfo.nCurrentIp & 0x0000ff00) >> 8);
			int nIp4 = (pstMVDevInfo->SpecialInfo.stGigEInfo.nCurrentIp & 0x000000ff);

			// ch:打印当前相机ip和用户自定义名字 | en:print current ip and user defined name
			printf("CurrentIp: %d.%d.%d.%d\n" , nIp1, nIp2, nIp3, nIp4);
			printf("UserDefinedName: %s\n\n" , pstMVDevInfo->SpecialInfo.stGigEInfo.chUserDefinedName);
		}
		else if (pstMVDevInfo->nTLayerType == MV_USB_DEVICE)
		{
			printf("UserDefinedName: %s\n", pstMVDevInfo->SpecialInfo.stUsb3VInfo.chUserDefinedName);
			printf("Serial Number: %s\n", pstMVDevInfo->SpecialInfo.stUsb3VInfo.chSerialNumber);
			printf("Device Number: %d\n\n", pstMVDevInfo->SpecialInfo.stUsb3VInfo.nDeviceNumber);
		}
		else if (pstMVDevInfo->nTLayerType == MV_GENTL_GIGE_DEVICE)
		{
			printf("UserDefinedName: %s\n", pstMVDevInfo->SpecialInfo.stGigEInfo.chUserDefinedName);
			printf("Serial Number: %s\n", pstMVDevInfo->SpecialInfo.stGigEInfo.chSerialNumber);
			printf("Model Name %s\n\n", pstMVDevInfo->SpecialInfo.stGigEInfo.chModelName);
		}
		else if (pstMVDevInfo->nTLayerType == MV_GENTL_CAMERALINK_DEVICE)
		{
			printf("UserDefinedName: %s\n", pstMVDevInfo->SpecialInfo.stCMLInfo.chUserDefinedName);
			printf("Serial Number: %s\n", pstMVDevInfo->SpecialInfo.stCMLInfo.chSerialNumber);
			printf("Model Name: %s\n\n", pstMVDevInfo->SpecialInfo.stCMLInfo.chModelName);
		}
		else if (pstMVDevInfo->nTLayerType == MV_GENTL_CXP_DEVICE)
		{
			printf("UserDefinedName: %s\n", pstMVDevInfo->SpecialInfo.stCXPInfo.chUserDefinedName);
			printf("Serial Number: %s\n", pstMVDevInfo->SpecialInfo.stCXPInfo.chSerialNumber);
			printf("Model Name: %s\n\n", pstMVDevInfo->SpecialInfo.stCXPInfo.chModelName);
		}
		else if (pstMVDevInfo->nTLayerType == MV_GENTL_XOF_DEVICE)
		{
			printf("UserDefinedName: %s\n", pstMVDevInfo->SpecialInfo.stXoFInfo.chUserDefinedName);
			printf("Serial Number: %s\n", pstMVDevInfo->SpecialInfo.stXoFInfo.chSerialNumber);
			printf("Model Name: %s\n\n", pstMVDevInfo->SpecialInfo.stXoFInfo.chModelName);
		}
		else
		{
			printf("Not support.\n");
		}
	}  

	return true;
}

bool PrintInterfaceInfo(MV_INTERFACE_INFO_LIST& stInterfaceInfoList)
{
	for (unsigned int i = 0; i < stInterfaceInfoList.nInterfaceNum; i++)
	{
		printf("[Interface %d]:\n", i);
		MV_INTERFACE_INFO* pstInterfaceInfo = stInterfaceInfoList.pInterfaceInfos[i];
		if (NULL == pstInterfaceInfo)
		{
			break;
		} 
		printf("Display name: %s\n",pstInterfaceInfo->chDisplayName);
		printf("Serial number: %s\n",pstInterfaceInfo->chSerialNumber);
		printf("model name: %s\n",pstInterfaceInfo->chModelName);
		printf("\n");            
	}

	return true;
}

void __stdcall EventCallBack(MV_EVENT_OUT_INFO * pEventInfo, void* pUser)
{
	if (pEventInfo)
	{
		__int64 nTimestamp = pEventInfo->nTimestampHigh;
		nTimestamp = (nTimestamp << 32) + pEventInfo->nTimestampLow;

		printf("EventName[%s], EventID[%u], Timestamp[%lld]\n", 
			pEventInfo->EventName, pEventInfo->nEventID,nTimestamp);
	}
}

void __stdcall ImageCallBackEx(unsigned char * pData, MV_FRAME_OUT_INFO_EX* pFrameInfo, void* pUser)
{
	/*if (pFrameInfo)
	{
		printf("Get One Frame: Width[%d], Height[%d], nFrameNum[%d]\n", 
			pFrameInfo->nExtendWidth, pFrameInfo->nExtendHeight, pFrameInfo->nFrameNum);
	}*/
}